Preview gif

HenryLeongStat opened the pull request:
Preview magic for .gif files.

BoPeng commented:
You do not have `image/png` in the multi-output case. This is needed for frontend that cannot handle gif, in that case `png` will be used. This is why I suggested that you create a `png` preview, and add `gif` preview when there are multiple output files.

You could use `shutil.rmtree` to remove the entire temporary directory instead of using three `os.remove`.

What is the purpose of [line 264](https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L264)? You do not really need `wand` if this is not needed.

BoPeng commented:
Also, read the comments on `tempfile.TemporaryDirectory` again

> On completion of the context or destruction of the temporary directory object the newly created temporary directory and all its contents are removed from the filesystem.

you are supposed to do

```
with tempfile.TemporaryDirectory(....) as tmp:
    do things with tmp.name as directory name
   # no need to shutil.rmtree.

HenryLeongStat commented:
Understand what you said for `with tempfile.TemporaryDirectory(....) as tmp:`, but I am confused about 
>You do not have image/png in the multi-output case. This is needed for frontend that cannot handle gif, in that case png will be used. This is why I suggested that you create a png preview, and add gif preview when there are multiple output files.

Because there are two cases:
1. Single output with `.png`
2. Multi output with `.gif`

What do you mean for "when frontend that cannot handle gif" ? 😖

BoPeng commented:
Because web is only one of the frontends of Jupyter kernel. You can use a terminal (ipython, qtconsole) etc to use it, and they may not be able to handle gif. In this case, png will be used. This is why many output of jupyter has text/html, text/plain etc ---- the frontend will use one of the results to present.

In our case, we should return png as base, and gif for frontends that can handle it.

HenryLeongStat commented:
Hmm.. I think I might understand... 
But how to specify image/png or image/gif as the return?

```
            try:
                #if len(pngFiles)==0:
                with open(outfile, 'rb') as content:
                    data = content.read()
                result = {'image/png': base64.b64encode(data).decode('ascii') }
                return result
                if len(pngFiles)!=0:
                    import imageio
                    pngFiles.sort(key=lambda x: int(x.split('.')[1]))
                    pngFiles.insert(0, join(tempDirectory, fileNameElement + '.png'))
                    images = []
                    for filename in pngFiles:
                        images.append(imageio.imread(join(tempDirectory, filename)))
                    gifName = fileNameElement + '.gif'
                    data = imageio.mimsave(join(tempDirectory, gifName), images, duration = 1)
                    with open(join(tempDirectory, gifName), 'rb') as f:
                        image = f.read()
                    image_data = base64.b64encode(image).decode('ascii')
                    remove(join(tempDirectory, gifName))
                    result['image/gif'] = base64.b64encode(image_data).decode('ascii')
                    return result
            except Exception as e:
                kernel.warn(e)
```

HenryLeongStat commented:
`return result['image/gif']` ?

HenryLeongStat commented:
Oh! I see where the problem is!

BoPeng commented:
I gave you the code yesterday

```
with open(outfile, 'rb') as content:
     data = content.read()
     result = { 'image/png':  base64.b64encode(data).decode('ascii') }
if there are more files:
    process
    result['image/gif'] = 
return result
```

HenryLeongStat commented:
Yes, that should be the same logic:
```
            try:
                #if len(pngFiles)==0:
                with open(outfile, 'rb') as content:
                    data = content.read()
                result = {'image/png': base64.b64encode(data).decode('ascii') }
                if len(pngFiles)!=0:
                    import imageio
                    pngFiles.sort(key=lambda x: int(x.split('.')[1]))
                    pngFiles.insert(0, join(tempDirectory, fileNameElement + '.png'))
                    images = []
                    for filename in pngFiles:
                        images.append(imageio.imread(join(tempDirectory, filename)))
                    gifName = fileNameElement + '.gif'
                    data = imageio.mimsave(join(tempDirectory, gifName), images, duration = 1)
                    with open(join(tempDirectory, gifName), 'rb') as f:
                        image = f.read()
                    image_data = base64.b64encode(image).decode('ascii')
                    remove(join(tempDirectory, gifName))
                    result['image/gif'] = image_data
                return result
            except Exception as e:
                kernel.warn(e)
```

However, even though there are multiple outputs, it can still be shown as `.png`, in that case, the first `{...}` will be shown.

BoPeng commented:
The second `result['image/gif'] =` should update the dictionary with an additional key, check the keys of `result` before you return it. Also, in the exception, return `result` with `png` so that the static version will be shown in case git-creating fails.

HenryLeongStat commented:
Understood! Should be working now!

BoPeng commented on src/sos/preview.py:
```diff
@@ -230,11 +230,45 @@ def preview_md(filename, kernel=None, style=None):
-        data = content.read()
-    return {'image/png': base64.b64encode(data).decode('ascii') }
+        fileNameElement = "sosDotFilesPng"
+        with tempfile.TemporaryDirectory() as tempDirectory:
```
No need for nested `with` since you already have `data` read from `outfile`, right?

HenryLeongStat replied:
You mean change it into `tempDirectory = tempfile.TemporaryDirectory().name`?
In this case, do I need to remove the TemporaryDirectory?

BoPeng commented on src/sos/preview.py:
```diff
@@ -230,11 +230,45 @@ def preview_md(filename, kernel=None, style=None):
+                    import imageio
+                    pngFiles.sort(key=lambda x: int(x.split('.')[1]))
+                    pngFiles.insert(0, join(tempDirectory, fileNameElement + '.png'))
+                    images = []
```
learn to do something like
```
images = [imageio.imread(.....x) for x in pngFiles]
```

HenryLeongStat replied:
Starting to find the list comprehensive style is very convenient!

HenryLeongStat commented:
`sudo apt-get update` fails? That isn't related to the code, is it?

BoPeng commented:
Are you sure both `Source` and `Render` need these two parameters?

```
        src = Source(dot.read(), filename = fileNameElement, directory = tempDirectory)
        src.format = 'png'
        outfile = src.render(filename = fileNameElement, directory = tempDirectory)
```

BoPeng commented:
`isfile(join(tempDirectory, f) `

not needed.

duplicated `  and fileNameElement in f`

HenryLeongStat commented:
Seems like we can't do 
```
with xxxx
with yyyy
```
, can we? Got an error from doing that.

The code here is `with X as x, Y as y`. According to [here](https://stackoverflow.com/questions/3024925/python-create-a-with-block-on-several-context-managers):

That's the same as:
```
with X as x:
    with Y as y:
```

In this case, that should be the same as [this patch](https://github.com/vatlab/SoS/pull/938/commits/08017b6164289c5ef342b090961eaba12b7fa4ed), right? Not very familiar with `with statement`...

BoPeng commented:
```
         with open(outfile, 'rb') as content:
                data = content.read()
            result = {'image/png': base64.b64encode(data).decode('ascii') }
```
should be put outside of `try/except`, otherwise if the block triggers an exception, `result` will not be created, and https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L269 will cause another exception.

BoPeng commented:
```
                remove(join(tempDirectory, gifName))
```
not needed.

https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L257

can write directly to an object, so you could potentially write to a `ByteIO` directly.

BoPeng commented:
https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L263

So you only return `gif`?? You should return both `png` and `gif`. If that is causing trouble, check why https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L72 worked.

HenryLeongStat commented:
>https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L263
So you only return gif?? You should return both png and gif. If that is causing trouble, check why https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L72 worked.

When I changed it into:
```
return {'image/gif': result['image/gif'],
             'image/png': result['image/png']}
```
It would show the **static** image instead of the **.gif**...

BoPeng commented:
Then check https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L77 and see if this is triggered instead of https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L72 when you preview `gif` file directly.

BoPeng commented:
It is possible that the browser chose `png` over `gif`, in which case we can only return one.

HenryLeongStat commented:
Same case it return a **static** image if I `%preview -n giphy.gif`! I think the reason why I can view a **.gif** yesterday is because I was using the older version of `SoS`?
<img width="302" alt="screen shot 2018-04-04 at 5 46 31 pm" src="https://user-images.githubusercontent.com/26916551/38338694-406b5b92-3830-11e8-8d18-636bed70f87d.png">

BoPeng commented:
That is because you did not have `wand` installed and https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L70 was triggered.

HenryLeongStat commented:
No, I fixed it this morning. 
<img width="398" alt="screen shot 2018-04-04 at 5 54 16 pm" src="https://user-images.githubusercontent.com/26916551/38338921-3eec3f74-3831-11e8-8a43-47425917b659.png">

I removed [L73](https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L73) and `%preview -n giphy.gif` works now.
<img width="288" alt="screen shot 2018-04-04 at 5 55 39 pm" src="https://user-images.githubusercontent.com/26916551/38338958-696c6a76-3831-11e8-82ab-48b63c3be9f2.png">

HenryLeongStat commented:
I thought the same, so I added `kernel.warn("Here")` before [L75](https://github.com/vatlab/SoS/blob/preview_gif/src/sos/preview.py#L75), but it didn't show up.

HenryLeongStat commented:
Just tried to `%preview -n file.gif` in [http://128.135.144.117:8000/](http://128.135.144.117:8000/). Still, it shows a **static** image instead of a **.gif** one. 
<img width="312" alt="screen shot 2018-04-04 at 6 02 57 pm" src="https://user-images.githubusercontent.com/26916551/38339206-bab1090e-3832-11e8-8ddd-d89fadeb6c38.png">

BoPeng commented:
you also need to fix gif preview. That is to say output gif without png for this format.

HenryLeongStat commented:
OK! Let me fix it now.

HenryLeongStat commented:
Done. I add a `if statement` specific for `.gif`. It works fine now:

<img width="595" alt="screen shot 2018-04-04 at 6 32 26 pm" src="https://user-images.githubusercontent.com/26916551/38339971-92a04afc-3836-11e8-8855-e39b571de5bf.png">
<img width="288" alt="screen shot 2018-04-04 at 6 32 35 pm" src="https://user-images.githubusercontent.com/26916551/38339974-94e1ad6a-3836-11e8-8782-86954c35e474.png">

HenryLeongStat commented:
Done.
